### PR TITLE
Added Oui for numeric entry

### PR DESCRIPTION
**Pull request by coloursofnoise**
Also added wrapper for OuiModOptionString for consistent naming

### CODE REVIEW

**Comment by coloursofnoise** [comment]
This has not been tested with a controller, and likely needs some work to be compatible

**Comment by max4805** [comment]
To try that out, I'm trying to make this screen usable with mod settings, by annotating an int or float property with `[SettingNumberInput]`.

**Comment by max4805** [comment]
The basics of controller support are here (you can type numbers, the keyboard grays out when no controller is plugged in like string input does), but here is a todo list:
- share a bit of input handling with the keyboard code, instead of concatenating all the time: allow decimals/negatives is ignored, and hitting "-" crashes the game
- selection disappears between "cancel" and "backspace"
- bonus points if we can gray out "-" when negatives are forbidden, and "." when decimals are forbidden.

**Comment by max4805** [comment]
Whoops.

image

**Comment by max4805** [comment]
Floating point precision is terrible, and typing 12345678 ends up setting the value to 12345680 instead.

**Comment by max4805** [comment]
Pushed what I did so far, I have to take a break now. Here is what is left to address:
- rounding issues with large numbers
- graying out - and . depending on options?
- the options on the bottom seem slightly misaligned

**Comment by max4805** [comment]
Grabbing this back! I was asked to test this out with a controller.

**Comment by max4805** [comment]
image
Hello I'm an annoying tester and I just found out that bruteforcing the centering of the keyboard doesn't quite work when switching to Chinese

**Comment by max4805** [comment]
Okay, I'm pretty much done now. Everything works as far as I can test.

File: Celeste.Mod.mm/Mod/UI/OuiModOptionString.cs [hunk 1]
```diff
@@ -423,7 +428,7 @@ public OuiModOptionString()
                 pos.Y += lineHeight + lineSpacing;
                 letterIndex++;
             }
-
+
```
Comment (0x0ade) [inline Celeste.Mod.mm/Mod/UI/OuiModOptionString.cs:431]:
```suggestion

```

File: Celeste.Mod.mm/Mod/UI/OuiNumberEntry.cs [hunk 2]
```diff
@@ -0,0 +1,508 @@
...
+        private Color selectColorB = Calc.HexToColor("FCFF59");
+        private Color disableColor = Color.DarkSlateBlue;
+
+        private Vector2 boxTopLeft {
+            get {
+                return Position + new Vector2((1920f - boxWidth) / 2f, 360f + (680f - boxHeight) / 2f);
+            }
+        }
+
+
```
Comment (0x0ade) [inline Celeste.Mod.mm/Mod/UI/OuiNumberEntry.cs:74]:
```suggestion

```

File: Celeste.Mod.mm/Mod/UI/OuiNumberEntry.cs [hunk 3]
```diff
@@ -0,0 +1,508 @@
...
+        /// <typeparam name="T">Oui to return to on exit</typeparam>
+        /// <param name="value">Initial value</param>
+        /// <param name="onValueChange">Action to be called when a new value is set.</param>
+        /// <param name="maxValueLength">The number of digits allowed, excluding "-" and "."</param>
+        /// <param name="allowDecimals">If decimal numbers should be allowed</param>
+        /// <param name="allowNegatives">If negative numbers should be allowed</param>
+        /// <returns></returns>
+        public OuiNumberEntry Init<T>(float value, Action<float> onValueChange, int maxValueLength = 6, bool allowDecimals = true, bool allowNegatives = true) where T : Oui {
+
+            _Value = StartingValue = value.ToString($"F{maxValueLength}").TrimEnd('0').TrimEnd('.');
```
Comment (0x0ade) [inline Celeste.Mod.mm/Mod/UI/OuiNumberEntry.cs:101]:
```suggestion
        public OuiNumberEntry Init<T>(float value, Action<float> onValueChange,
            int maxValueLength = 6, bool allowDecimals = true, bool allowNegatives = true) where T : Oui {
            _Value = StartingValue = value.ToString($"F{maxValueLength}").TrimEnd('0').TrimEnd('.');
```

File: Celeste.Mod.mm/Mod/UI/OuiNumberEntry.cs [hunk 4]
```diff
@@ -0,0 +1,508 @@
...
+            Vector2 posTo = new Vector2(0f, 1080f);
+            for (float t = 0f; t < 1f; t += Engine.DeltaTime * 2f) {
+                ease = 1f - Ease.CubeIn(t);
+                Position = posFrom + (posTo - posFrom) * Ease.CubeInOut(t);
+                yield return null;
+            }
+
+            Visible = false;
+        }
+        public void OnKeyboardInput(char c) {
```
Comment (0x0ade) [inline Celeste.Mod.mm/Mod/UI/OuiNumberEntry.cs:205]:
```suggestion
        }

        public void OnKeyboardInput(char c) {
```

**Pull request merged**